src/storage/src/region/writer.rs

```diff
@@ -270,7 +271,58 @@ impl RegionWriter {
 
         self.wait_flush().await?;
 
-        // TODO: canncel the compaction task
+        // TODO: cancel the compaction task
+
+        Ok(())
+    }
+
+    pub async fn on_drop<S: LogStore>(&self, drop_ctx: DropContext<'_, S>) -> Result<()> {
+        // 1. Acquires the write lock.
+        // 2. Close writer reject any potential writing.
+        // 3. Waits or cancels the flush job.
+        // 4. Add `RegionMetaAction::Remove` to recover from manifest in case of failure.
+        //    The main task is to restore the cleaning of sst files. If there is a failure
+        //    in the previous stops, it can be restored through the `Procedure` framework.
+        // 5. Mark all data obsolete in the WAL.
+        // 6. Delete the namespace of the region from the WAL.
+        // 7. Mark all SSTs deleted.
+        let mut inner = self.inner.lock().await;
+        inner.mark_closed();
+
+        if let Some(handle) = inner.flush_handle.take() {
+            handle.wait().await?;
+        }
+
+        let version_control = drop_ctx.version_control();
+
+        let _lock = self.version_mutex.lock().await;
+        let committed_sequence = version_control.committed_sequence();
+        let current_version = version_control.current();
+
+        let mut action_list =
+            RegionMetaActionList::with_action(RegionMetaAction::Remove(RegionRemove {
+                region_id: drop_ctx.shared.id,
+            }));
+
+        // Persist the meta action.
+        let prev_version = version_control.current_manifest_version();
+        action_list.set_prev_version(prev_version);
+
+        logging::debug!(
+            "Try to remove region {}, action_list: {:?}",
+            drop_ctx.shared.name(),
+            action_list
+        );
+
+        drop_ctx.manifest.update(action_list).await?;
```

Comment:
Should this manifest update be done via `write_edit_and_apply()` method?

Reply:
I don't think so, it doesn't modify meta and doesn't involve edit meta action. 
Maybe need to add `persist_manifest_version()` method.